src/mongocxx/exception/private/mongoc_error.hpp

```diff
@@ -0,0 +1,51 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <bsoncxx/document/value.hpp>
+
+#include <mongocxx/exception/error_category.hpp>
+#include <mongocxx/private/libmongoc.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+namespace exception {
+
+mongoc_error convert_mongoc_error(int code, int domain);
+std::error_code error_code_from_bson_error_t(::bson_error_t error);
+
+template <typename exception_type>
+void throw_exception(bsoncxx::document::value raw_server_error) {
```

Comment:
These are cool. The only problem with functions like these is that if, later, we wanted to adorn our exceptions with file/line information, we can't, because they will always actually be thrown from the same place.

I'm not sure I care, just pointing it out. It is one of the reasons that 'thing to help throw an exception' is frequently implemented a a MACRO rather than a function.

Not suggesting we change it though.